crypto/thread/thread_external.c

```diff
@@ -0,0 +1,329 @@
+
+#if defined(OPENSSL_THREADS)
+# include <openssl/crypto.h>
+
+# include "e_os.h"
+
+#  if defined(_WIN32)
+#   include <windows.h>
+#  endif
+
+# include <openssl/async.h>
+
+# include "thread.h"
+# include "thread_external.h"
+# include "internal/list.h"
+
+/* The following is required to keep ABI compat with CRYPTO_THREAD_ROUTINE
+ * as defined in openssl/crypto.h */
+# if !defined(CALLBACK)
+#  if defined(_WIN32) || defined(_STDCALL_SUPPORTED)
+#   define CALLBACK __stdcall
+#  else
+#   define CALLBACK
+#  endif
+# elif !defined(_WIN32) && !defined(_STDCALL_SUPPORTED)
+#  undef CALLBACK
+# endif
+
+volatile int CRYPTO_THREAD_EXTERN_enabled = 0;
+
+# ifdef OPENSSL_NO_EXTERN_THREAD
+
+int CRYPTO_THREAD_EXTERN_enable(CRYPTO_SIGNAL** props)
+{
+    return 0;
+}
+
+int CRYPTO_THREAD_EXTERN_disable()
+{
+    return 1;
+}
+
+# else /* ! OPENSSL_NO_EXTERN_THREAD */
+
+static struct list    CRYPTO_THREAD_EXTERN_task_queue;
+static struct list    CRYPTO_THREAD_EXTERN_task_done;
+static CRYPTO_MUTEX   CRYPTO_THREAD_EXTERN_task_lock;
+static CRYPTO_CONDVAR CRYPTO_THREAD_EXTERN_task_cond_create;
+static CRYPTO_CONDVAR CRYPTO_THREAD_EXTERN_task_cond_finish;
+
+int CRYPTO_THREAD_EXTERN_enable(CRYPTO_SIGNAL** props)
+{
+    if (CRYPTO_THREAD_EXTERN_enabled == 1)
+        return 1;
+
+    if (props != NULL && CRYPTO_SIGNAL_block_set(props) != 1)
+        goto fail;
+
+    list_init(&CRYPTO_THREAD_EXTERN_task_queue);
+    list_init(&CRYPTO_THREAD_EXTERN_task_done);
+
+    CRYPTO_THREAD_EXTERN_task_lock = CRYPTO_MUTEX_create();
+    CRYPTO_THREAD_EXTERN_task_cond_create = CRYPTO_CONDVAR_create();
+    CRYPTO_THREAD_EXTERN_task_cond_finish = CRYPTO_CONDVAR_create();
+
+    if (CRYPTO_MUTEX_init(CRYPTO_THREAD_EXTERN_task_lock) == 0)
+        goto fail;
+
+    if (CRYPTO_CONDVAR_init(CRYPTO_THREAD_EXTERN_task_cond_create) == 0)
+        goto fail;
+
+    if (CRYPTO_CONDVAR_init(CRYPTO_THREAD_EXTERN_task_cond_finish) == 0)
+        goto fail;
+
+    CRYPTO_mem_barrier();
+
+    CRYPTO_THREAD_EXTERN_enabled = 1;
+    return 1;
+
+fail:
+    OPENSSL_free(CRYPTO_THREAD_EXTERN_task_lock);
+    OPENSSL_free(CRYPTO_THREAD_EXTERN_task_cond_create);
+    OPENSSL_free(CRYPTO_THREAD_EXTERN_task_cond_finish);
+
+    CRYPTO_THREAD_EXTERN_disable();
+    return 0;
+}
+
+int CRYPTO_THREAD_EXTERN_disable(void)
+{
+    if (CRYPTO_THREAD_EXTERN_enabled) {
+        CRYPTO_MUTEX_destroy(&CRYPTO_THREAD_EXTERN_task_lock);
+        CRYPTO_CONDVAR_destroy(&CRYPTO_THREAD_EXTERN_task_cond_create);
+        CRYPTO_CONDVAR_destroy(&CRYPTO_THREAD_EXTERN_task_cond_finish);
+    }
+
+    CRYPTO_THREAD_EXTERN_enabled = 0;
+    return 1;
+}
+
+# endif
+
+static int CRYPTO_THREAD_EXTERN_job(void* arg)
+{
+    ASYNC_JOB * currjob;
+    CRYPTO_THREAD_TASK* task;
+
+    if (arg == NULL)
+        return 0;
+
+    task = (CRYPTO_THREAD_TASK*)arg;
+
+    if ((currjob = ASYNC_get_current_job()) == NULL)
+        return 0;
+
+    ASYNC_block_pause();
+    task->retval = task->task(task->data);
```

Comment:
I am confused by what you are attempting to achieve with all this ASYNC code. You seem to spend a lot of effort setting up an ASYNC job, only to then block pauses during it, which defeats the object. Why do you need to do anything with ASYNC at all?